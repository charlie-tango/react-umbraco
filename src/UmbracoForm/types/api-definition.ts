export interface paths {
  "/umbraco/forms/api/v1/definitions/{id}": {
    get: {
      parameters: {
        query?: {
          /** @description The Id of the content page on which the form is hosted. */
          contentId?: string;
          /** @description The culture code for the form's localization context. */
          culture?: string;
        };
        path: {
          /** @description The form's Id. */
          id: string;
        };
      };
      responses: {
        /** @description Success */
        200: {
          content: {
            "application/json": components["schemas"]["FormDto"];
          };
        };
        /** @description Bad Request */
        400: {
          content: {
            "application/json":
              | components["schemas"]["ProblemDetails"]
              | components["schemas"]["HttpValidationProblemDetails"];
          };
        };
        /** @description Not Found */
        404: {
          content: {
            "application/json":
              | components["schemas"]["ProblemDetails"]
              | components["schemas"]["HttpValidationProblemDetails"];
          };
        };
      };
    };
  };
  "/umbraco/forms/api/v1/entries/{id}": {
    post: {
      parameters: {
        path: {
          /** @description The form's Id. */
          id: string;
        };
      };
      requestBody?: {
        content: {
          "application/json": components["schemas"]["FormEntryDto"];
        };
      };
      responses: {
        /** @description Accepted */
        202: {
          content: never;
        };
        /** @description Bad Request */
        400: {
          content: {
            "application/json":
              | components["schemas"]["ProblemDetails"]
              | components["schemas"]["HttpValidationProblemDetails"];
          };
        };
        /** @description Not Found */
        404: {
          content: {
            "application/json":
              | components["schemas"]["ProblemDetails"]
              | components["schemas"]["HttpValidationProblemDetails"];
          };
        };
        /** @description Client Error */
        422: {
          content: {
            "application/json":
              | components["schemas"]["ProblemDetails"]
              | components["schemas"]["HttpValidationProblemDetails"];
          };
        };
      };
    };
  };
}

export type webhooks = Record<string, never>;

export interface components {
  schemas: {
    /** @enum {string} */
    FieldConditionActionType: "Show" | "Hide";
    /** @enum {string} */
    FieldConditionLogicType: "All" | "Any";
    /** @enum {string} */
    FieldConditionRuleOperator:
      | "Is"
      | "IsNot"
      | "GreaterThen"
      | "LessThen"
      | "Contains"
      | "ContainsIgnoreCase"
      | "StartsWith"
      | "StartsWithIgnoreCase"
      | "EndsWith"
      | "EndsWithIgnoreCase"
      | "NotContains"
      | "NotContainsIgnoreCase"
      | "NotStartsWith"
      | "NotStartsWithIgnoreCase"
      | "NotEndsWith"
      | "NotEndsWithIgnoreCase";
    FormConditionDto: {
      actionType?: components["schemas"]["FieldConditionActionType"];
      logicType?: components["schemas"]["FieldConditionLogicType"];
      rules?: components["schemas"]["FormConditionRuleDto"][];
    };
    FormConditionRuleDto: {
      field?: string;
      operator?: components["schemas"]["FieldConditionRuleOperator"];
      value?: string;
    };
    FormDto: {
      /** Format: uuid */
      id?: string;
      name?: string;
      indicator?: string;
      cssClass?: string;
      nextLabel?: string;
      previousLabel?: string;
      submitLabel?: string;
      disableDefaultStylesheet?: boolean;
      fieldIndicationType?: components["schemas"]["FormFieldIndication"];
      hideFieldValidation?: boolean;
      messageOnSubmit?: string;
      messageOnSubmitIsHtml?: boolean;
      showValidationSummary?: boolean;
      /** Format: uuid */
      gotoPageOnSubmit?: string;
      gotoPageOnSubmitRoute?: components["schemas"]["IApiContentRouteModel"];
      pages?: components["schemas"]["FormPageDto"][];
    };
    FormEntryDto: {
      values?: {
        [key: string]: string[];
      };
      contentId?: string;
      culture?: string;
    };
    FormFieldDto: {
      /** Format: uuid */
      id?: string;
      caption?: string;
      helpText?: string;
      placeholder?: string;
      cssClass?: string;
      alias?: string;
      required?: boolean;
      requiredErrorMessage?: string;
      pattern?: string;
      patternInvalidErrorMessage?: string;
      condition?: components["schemas"]["FormConditionDto"];
      fileUploadOptions?: components["schemas"]["FormFileUploadOptionsDto"];
      preValues?: components["schemas"]["FormFieldPrevalueDto"][];
      settings?: {
        [key: string]: string;
      };
      type?: components["schemas"]["FormFieldTypeDto"];
    };
    /** @enum {string} */
    FormFieldIndication:
      | "NoIndicator"
      | "MarkMandatoryFields"
      | "MarkOptionalFields";
    FormFieldPrevalueDto: {
      value?: string;
      caption?: string;
    };
    FormFieldTypeDto: {
      /** Format: uuid */
      id?: string;
      name?: string;
      supportsPreValues?: boolean;
      supportsUploadTypes?: boolean;
      renderInputType?: string;
    };
    FormFieldsetColumnDto: {
      caption?: string;
      /** Format: int32 */
      width?: number;
      fields?: components["schemas"]["FormFieldDto"][];
    };
    FormFieldsetDto: {
      /** Format: uuid */
      id?: string;
      caption?: string;
      condition?: components["schemas"]["FormConditionDto"];
      columns?: components["schemas"]["FormFieldsetColumnDto"][];
    };
    FormFileUploadOptionsDto: {
      allowAllUploadExtensions?: boolean;
      allowedUploadExtensions?: string[];
      allowMultipleFileUploads?: boolean;
    };
    FormPageDto: {
      caption?: string;
      condition?: components["schemas"]["FormConditionDto"];
      fieldsets?: components["schemas"]["FormFieldsetDto"][];
    };
    HttpValidationProblemDetails: {
      errors?: {
        [key: string]: string[];
      };
      [key: string]: unknown;
    } & components["schemas"]["ProblemDetails"];
    IApiContentRouteModel: {
      path: string;
      startItem?: components["schemas"]["IApiContentStartItemModel"];
    };
    IApiContentStartItemModel: {
      /** Format: uuid */
      id: string;
      path: string;
    };
    ProblemDetails: {
      type?: string;
      title?: string;
      /** Format: int32 */
      status?: number;
      detail?: string;
      instance?: string;
      [key: string]: unknown;
    };
  };
  responses: never;
  parameters: never;
  requestBodies: never;
  headers: never;
  pathItems: never;
}

export type $defs = Record<string, never>;

export type external = Record<string, never>;

export type operations = Record<string, never>;
