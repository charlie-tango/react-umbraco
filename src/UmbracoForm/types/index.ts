import type { ZodSchema, ZodTypeAny } from "zod";
import type { components } from "./api-definition";

/** Enum of default form field type ids
 * @see https://docs.umbraco.com/umbraco-forms/developer/configuration/type-details#field-types */
export enum DefaultFieldType {
  Checkbox = "d5c0c390-ae9a-11de-a69e-666455d89593",
  DataConsent = "a72c9df9-3847-47cf-afb8-b86773fd12cd",
  Date = "f8b4c3b8-af28-11de-9dd8-ef5956d89593",
  DropdownList = "0dd29d42-a6a5-11de-a2f2-222256d89593",
  FileUpload = "84a17cf8-b711-46a6-9840-0e4a072ad000",
  LongAnswer = "023f09ac-1445-4bcb-b8fa-ab49f33bd046",
  HiddenField = "da206cae-1c52-434e-b21a-4a7c198af877",
  MultipleChoice = "fab43f20-a6bf-11de-a28f-9b5755d89593",
  Password = "fb37bc60-d41e-11de-aeae-37c155d89593",
  Recaptcha2 = "b69deaeb-ed75-4dc9-bfb8-d036bf9d3730",
  RecaptchaV3WithScore = "663aa19b-423d-4f38-a1d6-c840c926ef86",
  RichText = "1f8d45f8-76e6-4550-a0f5-9637b8454619",
  SingleChoice = "903df9b0-a78c-11de-9fc1-db7a56d89593",
  ShortAnswer = "3f92e01b-29e2-4a30-bf33-9df5580ed52c",
  TitleAndDescription = "e3fbf6c4-f46c-495e-aff8-4b3c227b4a98",
}

export interface DefaultFieldTypeNameMap {
  [DefaultFieldType.Checkbox]: "Checkbox";
  [DefaultFieldType.DataConsent]: "Data consent";
  [DefaultFieldType.Date]: "Date";
  [DefaultFieldType.DropdownList]: "Dropdown";
  [DefaultFieldType.FileUpload]: "File upload";
  [DefaultFieldType.LongAnswer]: "Long answer";
  [DefaultFieldType.HiddenField]: "Hidden field";
  [DefaultFieldType.ShortAnswer]: "Short answer";
  [DefaultFieldType.MultipleChoice]: "Multiple choice";
  [DefaultFieldType.Recaptcha2]: "Recaptcha2";
  [DefaultFieldType.RecaptchaV3WithScore]: "Recaptcha v3 with score";
  [DefaultFieldType.Password]: "Password";
  [DefaultFieldType.RichText]: "Rich text";
  [DefaultFieldType.SingleChoice]: "Single choice";
  [DefaultFieldType.TitleAndDescription]: "Title and description";
}

/** Field type settings
 * @see https://docs.umbraco.com/umbraco-forms/developer/configuration/type-details#field-types */
export interface FieldSettings {
  [DefaultFieldType.ShortAnswer]: {
    defaultValue: string;
    placeholder: string;
    showLabel: string;
    maximumLength: string;
    fieldType: string;
    autocompleteAttribute: string;
  };
  [DefaultFieldType.LongAnswer]: {
    defaultValue: string;
    placeholder: string;
    showLabel: string;
    autocompleteAttribute: string;
    numberOfRows: string;
    maximumLength: string;
  };
  [DefaultFieldType.HiddenField]: {
    defaultValue: string;
  };
  [DefaultFieldType.Checkbox]: {
    caption: string;
    defaultValue: string;
    showLabel: string;
  };
  [DefaultFieldType.DropdownList]: {
    defaultValue: string;
    allowMultipleSelections: string;
    showLabel: string;
    autocompleteAttribute: string;
    selectPrompt: string;
  };
  [DefaultFieldType.MultipleChoice]: {
    defaultValue: string;
    showLabel: string;
    preValues: Array<unknown>;
  };
  [DefaultFieldType.DataConsent]: {
    acceptCopy: string;
    showLabel: string;
  };
  [DefaultFieldType.FileUpload]: {
    selectFilesListHeading: string;
  };
  [DefaultFieldType.Recaptcha2]: {
    theme: string;
    size: string;
    errorMessage: string;
  };
  [DefaultFieldType.RecaptchaV3WithScore]: {
    scoreThreshold: string;
    errorMessage: string;
    saveScore: string;
  };
  [DefaultFieldType.Date]: {
    placeholder: string;
  };
  [DefaultFieldType.Password]: {
    placeholder: string;
  };
  [DefaultFieldType.RichText]: {
    showLabel: string;
    html: string;
  };
  [DefaultFieldType.SingleChoice]: {
    defaultValue: string;
    showLabel: string;
  };
  [DefaultFieldType.TitleAndDescription]: {
    captionTag: string;
    caption: string;
    bodyText: string;
    showLabel: string;
  };
}

/** Form definition
 * @see https://docs.umbraco.com/umbraco-forms/editor/creating-a-form/form-settings */
export interface FormDto
  extends Omit<components["schemas"]["FormDto"], "pages"> {
  pages?: FormPageDto[];
}

/** Form page
 * @see https://docs.umbraco.com/umbraco-forms/editor/creating-a-form#form-pages */
export interface FormPageDto
  extends Omit<components["schemas"]["FormPageDto"], "fieldsets"> {
  fieldsets?: FormFieldsetDto[];
}

/** Form groups
 * @see https://docs.umbraco.com/umbraco-forms/editor/creating-a-form#form-groups */
export interface FormFieldsetDto
  extends Omit<components["schemas"]["FormFieldsetDto"], "columns"> {
  columns?: FormFieldsetColumnDto[];
}

/** Form fieldset column
 * @see https://docs.umbraco.com/umbraco-forms/editor/creating-a-form#form-columns */
export interface FormFieldsetColumnDto
  extends Omit<components["schemas"]["FormFieldsetColumnDto"], "fields"> {
  fields?: FormFieldDto[];
}

/** Form field type
 * @see https://docs.umbraco.com/umbraco-forms/developer/configuration/type-details#field-types */
export interface FormFieldTypeDto
  extends Omit<components["schemas"]["FormFieldTypeDto"], "name"> {
  name: DefaultFieldTypeNameMap[DefaultFieldType];
  id: DefaultFieldType;
}

/** Form field
 * @see https://docs.umbraco.com/umbraco-forms/developer/configuration/type-details#field-types */
export interface FormFieldDto
  extends Omit<components["schemas"]["FormFieldDto"], "type" | "settings"> {
  type: FormFieldTypeDto;
  settings: FieldSettings[DefaultFieldType];
}

/** Form condition
 * @see https://docs.umbraco.com/umbraco-forms/editor/creating-a-form/conditional-logic */
export type FormConditionDto = components["schemas"]["FormConditionDto"];

/** Form condition rule
 * @see https://docs.umbraco.com/umbraco-forms/editor/creating-a-form/conditional-logic */
export type FormConditionRuleDto =
  components["schemas"]["FormConditionRuleDto"];

/** Field condition action type
 * @see https://docs.umbraco.com/umbraco-forms/editor/creating-a-form/conditional-logic#action-and-logic-types */
export type FieldConditionActionType =
  components["schemas"]["FieldConditionActionType"];

/** Field condition logic type
 * @see https://docs.umbraco.com/umbraco-forms/editor/creating-a-form/conditional-logic#action-and-logic-types */
export type FieldConditionLogicType =
  components["schemas"]["FieldConditionLogicType"];

/** Field condition rule operator
 * @see https://docs.umbraco.com/umbraco-forms/editor/creating-a-form/conditional-logic#adding-a-new-condition */
export type FieldConditionRuleOperator =
  components["schemas"]["FieldConditionRuleOperator"];

/** Data transfer object with conditions, includes pages, fieldsets and fields
 * @see https://docs.umbraco.com/umbraco-forms/editor/creating-a-form/conditional-logic */
export type DtoWithCondition = FormPageDto | FormFieldsetDto | FormFieldDto;

/** Configuration for the UmbracoForm component instance */
export type UmbracoFormConfig = {
  /** Custom schema for form validation; defaults to the umbracoFormToZod implementation */
  schema: ZodSchema;
  /** Optional custom function to map form fields to Zod types, useful for handling custom field types */
  mapCustomFieldToZodType?: MapFormFieldToZodFn;
} & (
  | {
      /** Flag indicating if client-side validation should be performed; defaults to `false` */
      shouldValidate: true;
      /** Flag indicating if native browser validation should be used alongside client-side validation; defaults to `false` */
      shouldUseNativeValidation?: boolean;
      /** Flag indicating if the form should be validated on change, blur, or submit; defaults to `onSubmit` */
      validateMode?: "onBlur" | "onSubmit" | "onChange" | "all";
      /** Re-validate the form on change, blur, or submit; defaults to `onBlur` */
      reValidateMode?: "onChange" | "onBlur" | "onSubmit";
    }
  | {
      shouldValidate?: false;
      shouldUseNativeValidation?: never;
      validateMode?: never;
      reValidateMode?: never;
    }
);

export type MapFormFieldToZodFn = (field?: FormFieldDto) => ZodTypeAny;
