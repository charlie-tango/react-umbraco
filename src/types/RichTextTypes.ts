import type { Overwrite } from "../utils/helper-types";

export function isRootElement(data: RichTextElementModel | undefined): data is {
  tag: "#root";
  attributes?: Record<string, unknown>;
  elements: RichTextElementModel[];
  blocks?: Array<UmbracoBlockContext>;
} {
  return !!data && data.tag === "#root";
}

export function isTextElement(
  data: RichTextElementModel,
): data is { tag: "#text"; text: string } {
  return data.tag === "#text";
}

export function isCommentElement(
  data: RichTextElementModel,
): data is { tag: "#comment"; text: string } {
  return data.tag === "#comment";
}

export function isUmbracoBlock(data: RichTextElementModel): data is {
  tag: string;
  attributes: {
    "content-id": string;
  };
  elements: RichTextElementModel[];
} {
  return data.tag === "umb-rte-block" || data.tag === "umb-rte-block-inline";
}

export function isHtmlElement(data: RichTextElementModel): data is {
  tag: keyof HTMLElementTagNameMap;
  attributes: Record<string, unknown> & { route?: RouteAttributes };
  elements: RichTextElementModel[];
} {
  return "elements" in data;
}

interface BaseBlockItemModel {
  content?: {
    id: string;
    properties: {
      [key: string]: unknown;
    };
  };
  settings?: {
    id: string;
    properties: {
      [key: string]: unknown;
    };
  };
}

/**
 * Override the block item model with Umbraco specific properties.
 * This way you can get the full type safety of the Umbraco API you are using.
 *
 * **react-umbraco.d.ts**
 * ```ts
 * import { components } from '@/openapi/umbraco';
 *
 * // Define the intermediate interface
 * type ApiBlockItemModel = components['schemas']['ApiBlockItemModel'];
 *
 * declare module '@charlietango/react-umbraco' {
 *   interface UmbracoBlockItemModel extends ApiBlockItemModel {}
 * }
 * ```
 */
export interface UmbracoBlockItemModel {
  _overrideImplementation?: never;
}

export type UmbracoBlockContext = Overwrite<
  BaseBlockItemModel,
  Omit<UmbracoBlockItemModel, "_overrideImplementation">
>;

export interface RouteAttributes {
  path: string;
  startItem: {
    id: string;
    path: string;
  };
}

/**
 * The possible values for the rich text node from Umbraco.
 */
export type RichTextElementModel =
  | {
      tag: "#root";
      attributes?: Record<string, unknown>;
      elements?: RichTextElementModel[];
      blocks?: Array<UmbracoBlockContext>;
    }
  | {
      tag: "#text";
      text: string;
    }
  | {
      tag: "#comment";
      text: string;
    }
  | {
      tag: "umb-rte-block" | "umb-rte-block-inline";
      attributes: {
        "content-id": string;
      };
      elements: RichTextElementModel[];
    }
  | {
      tag: keyof HTMLElementTagNameMap;
      attributes: Record<string, unknown> & { route?: RouteAttributes };
      elements?: RichTextElementModel[];
    }
  | {
      tag: string;
      attributes?: Record<string, unknown> & { route?: RouteAttributes };
      elements?: RichTextElementModel[];
      blocks?: Array<UmbracoBlockContext>;
    };
