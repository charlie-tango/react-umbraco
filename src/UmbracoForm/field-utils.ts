import { match } from "ts-pattern";
import type { z } from "zod";
import { isConditionFulfilled } from "./conditions";
import type {
  DtoWithCondition,
  FieldSettings,
  FormDto,
  FormFieldDto,
  FormPageDto,
  UmbracoFormConfig,
} from "./types";
import { DefaultFieldType } from "./types";
import { type MapFormFieldToZodFn, getIssueId } from "./umbraco-form-to-zod";

const cachedFieldsById = new WeakMap<FormDto, Map<string, FormFieldDto>>();
const cachedFieldsByAlias = new WeakMap<FormDto, Map<string, FormFieldDto>>();
const cachedFieldsByPage = new WeakMap<FormPageDto, FormFieldDto[]>();

/**
 * Retrieves all fields from a given form.
 * If the form's fields are already cached, it returns them from the cache.
 * Otherwise, it flattens the fields from all pages, fieldsets, and columns,
 * caches them, and then returns the flattened list of fields.
 *
 * @param {FormDto} form - The form definition to retrieve fields from.
 * @returns {FormFieldDto[]} An array of form fields from the form definition.
 */
export function getAllFields(form: FormDto): FormFieldDto[] {
  if (cachedFieldsById.has(form)) {
    return Array.from(cachedFieldsById.get(form)?.values() ?? []);
  }
  const flattenedFields = form?.pages
    ?.flatMap((page) =>
      page?.fieldsets?.flatMap((fieldset) =>
        fieldset?.columns
          ?.flatMap((column) => column.fields)
          ?.map((field, index, array) => {
            if (index === 0) {
              // cache fields by page
              cachedFieldsByPage.set(page, array as FormFieldDto[]);
            }
            return field;
          }),
      ),
    )
    .filter(Boolean) as FormFieldDto[];

  /// cache flattened fields by id and alias
  const idMap = new Map<string, FormFieldDto>();
  const aliasMap = new Map<string, FormFieldDto>();

  for (const field of flattenedFields) {
    if (field.id) idMap.set(field.id, field);
    if (field.alias) aliasMap.set(field.alias, field);
  }
  cachedFieldsById.set(form, idMap);
  cachedFieldsByAlias.set(form, aliasMap);

  return flattenedFields;
}

/**
 * Retrieves a field from a form by its id.
 * If the field is cached, it returns the cached field.
 * Otherwise, it finds the field by iterating over all fields.
 *
 * @param {FormDto} form - The form data transfer object.
 * @param {string} id - The id of the field to retrieve.
 * @returns {FieldDto | undefined} The field with the specified alias, or undefined if not found.
 */
export function getFieldById(
  form: FormDto,
  id?: string,
): FormFieldDto | undefined {
  if (!id) return undefined;
  if (cachedFieldsById.has(form)) {
    return cachedFieldsById.get(form)?.get(id);
  }
  return getAllFields(form)?.find((field) => field?.id === id);
}

/**
 * Retrieves a field from a form by its alias.
 * If the field is cached, it returns the cached field.
 * Otherwise, it finds the field by iterating over all fields.
 *
 * @param {FormDto} form - The form data transfer object.
 * @param {string} alias - The alias of the field to retrieve.
 * @returns {FieldDto | undefined} The field with the specified alias, or undefined if not found.
 */
export function getFieldByAlias(
  form: FormDto,
  alias: string,
): FormFieldDto | undefined {
  if (cachedFieldsByAlias.has(form)) {
    return cachedFieldsByAlias.get(form)?.get(alias);
  }
  return getAllFields(form)?.find((field) => field?.alias === alias);
}

/**
 * Retrieves all fields on a given page
 * @param {FormPageDto} [page] - The page object containing fieldsets to process.
 * @returns {FormFieldDto[]} An array of form field objects.
 */
export function getAllFieldsOnPage(page?: FormPageDto): FormFieldDto[] {
  if (page && cachedFieldsByPage.has(page)) {
    return cachedFieldsByPage.get(page) ?? [];
  }
  return (
    (page?.fieldsets
      ?.flatMap((fieldset) => fieldset?.columns)
      ?.flatMap((column) => column?.fields)
      .filter(Boolean) as FormFieldDto[]) ?? []
  );
}

/**
 * Retrieves the form field associated with a given Zod validation issue.
 *
 * @param {FormDto} form - The form definition.
 * @param {z.ZodIssue} issue - The Zod issue containing the error details.
 * @returns The form field corresponding to the issue's path.
 */
export function getFieldByZodIssue(form: FormDto, issue: z.ZodIssue) {
  const alias = issue.path.join(".");
  return getFieldByAlias(form, alias);
}
/**
 * Filters form fields based on conditions.
 *
 * @param {FormDto} form - The form definition.
 * @param {Record<string, unknown>} data - The data to check against the conditions.
 * @param {MapFormFieldToZodFn} [mapCustomFieldToZodType] - Optional mapping of custom fields to Zod validation schema.
 * @returns {FormFieldDto[]} An array of form fields that meet the specified conditions.
 */
export function filterFieldsByConditions(
  form: FormDto,
  data: Record<string, unknown>,
  mapCustomFieldToZodType?: MapFormFieldToZodFn,
): FormFieldDto[] {
  const checkCondition = (dto?: DtoWithCondition) =>
    dto
      ? isConditionFulfilled(dto, form, data, mapCustomFieldToZodType)
      : false;

  return form?.pages
    ?.filter(checkCondition)
    ?.flatMap((page) => page?.fieldsets)
    ?.filter(checkCondition)
    ?.flatMap((fieldset) => fieldset?.columns)
    ?.flatMap((column) => column?.fields)
    ?.filter(checkCondition) as FormFieldDto[];
}

type CommonAttributes = React.InputHTMLAttributes<HTMLInputElement> &
  React.TextareaHTMLAttributes<HTMLTextAreaElement> &
  React.SelectHTMLAttributes<HTMLSelectElement>;

export function getAttributesForFieldType(
  field: FormFieldDto,
  issues: z.ZodIssue[] | undefined,
  form: FormDto,
  config: UmbracoFormConfig,
) {
  const { hideFieldValidation, showValidationSummary } = form;
  const { shouldValidate, shouldUseNativeValidation } = config;
  const hasIssues = issues && issues?.length > 0;

  const rendersOnlySummary =
    showValidationSummary === true && hideFieldValidation === true;
  const rendersOnlyLabelError =
    showValidationSummary === false && hideFieldValidation === false;

  const errorsAreHidden =
    showValidationSummary === false && hideFieldValidation === true;

  const ariaInvalid = "aria-invalid";
  const ariaErrorMessage = "aria-errormessage";
  const ariaDescribedBy = "aria-describedby";

  const commonAttributes: CommonAttributes = {
    name: field.alias,
    id: field.id,
    required:
      shouldValidate && shouldUseNativeValidation && field.required
        ? field.required
        : undefined,
    [ariaInvalid]: shouldValidate ? hasIssues : undefined,
    [ariaErrorMessage]:
      // only adds error message if no error elements are present in the DOM
      shouldValidate && hasIssues && errorsAreHidden
        ? issues[0].message
        : undefined,
    // show only aria-describedby if error elements are present in the DOM
    [ariaDescribedBy]:
      shouldValidate &&
      hasIssues &&
      (rendersOnlySummary || rendersOnlyLabelError)
        ? getIssueId(field, issues[0])
        : undefined,
  };

  const defaultValue =
    "defaultValue" in field.settings ? field.settings.defaultValue : undefined;

  const textAttributes = match(field?.type?.id)
    .with(DefaultFieldType.ShortAnswer, DefaultFieldType.LongAnswer, (id) => {
      const settings = field?.settings as FieldSettings[typeof id];
      return {
        autoComplete: settings?.autocompleteAttribute || undefined,
        placeholder: field.placeholder || undefined,
        pattern:
          shouldValidate && shouldUseNativeValidation && field.pattern
            ? field.pattern
            : undefined,
        maxLength:
          shouldValidate && shouldUseNativeValidation && settings?.maximumLength
            ? Number.parseInt(settings?.maximumLength)
            : undefined,
      };
    })
    .otherwise(() => {});

  return match(field?.type?.id.toLowerCase())
    .with(DefaultFieldType.ShortAnswer, (id) => {
      const settings = field?.settings as FieldSettings[typeof id];
      return {
        type: settings?.fieldType || "text",
        ...commonAttributes,
        defaultValue,
        ...textAttributes,
      } satisfies React.InputHTMLAttributes<HTMLInputElement>;
    })
    .with(DefaultFieldType.LongAnswer, (id) => {
      const settings = field?.settings as FieldSettings[typeof id];
      return {
        defaultValue,
        ...textAttributes,
        ...commonAttributes,
        rows: settings?.numberOfRows
          ? Number.parseInt(settings.numberOfRows)
          : undefined,
      } satisfies React.TextareaHTMLAttributes<HTMLTextAreaElement>;
    })
    .with(DefaultFieldType.MultipleChoice, () => ({
      type: "radio",
      ...commonAttributes,
    }))
    .with(
      DefaultFieldType.Checkbox,
      DefaultFieldType.DataConsent,
      () =>
        ({
          type: "checkbox",
          defaultChecked: !!defaultValue,
          ...commonAttributes,
        }) satisfies React.InputHTMLAttributes<HTMLInputElement>,
    )
    .with(
      DefaultFieldType.Recaptcha2,
      DefaultFieldType.RecaptchaV3WithScore,
      () =>
        ({
          type: "hidden",
          ...commonAttributes,
        }) satisfies React.InputHTMLAttributes<HTMLInputElement>,
    )
    .with(DefaultFieldType.DropdownList, (id) => {
      const settings = field?.settings as FieldSettings[typeof id];
      return {
        defaultValue,
        ...commonAttributes,
        multiple: !!settings?.allowMultipleSelections ?? false,
      } satisfies React.SelectHTMLAttributes<HTMLSelectElement>;
    })
    .with(DefaultFieldType.Date, () => ({
      type: "date",
      ...commonAttributes,
    }))
    .with(DefaultFieldType.Password, () => ({
      type: "password",
      ...commonAttributes,
    }))
    .with(DefaultFieldType.RichText, () => ({
      type: "textarea",
      ...commonAttributes,
    }))
    .with(
      DefaultFieldType.FileUpload,
      () =>
        ({
          type: "file",
          ...commonAttributes,
          accept: field?.fileUploadOptions?.allowedUploadExtensions?.join(","),
        }) satisfies React.InputHTMLAttributes<HTMLInputElement>,
    )
    .otherwise(() => commonAttributes);
}
