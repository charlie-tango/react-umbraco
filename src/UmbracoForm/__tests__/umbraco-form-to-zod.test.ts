import { z } from "zod";
import { getAllFields } from "../field-utils";
import { DefaultFieldType } from "../types";
import type { FormDto, FormFieldDto } from "../types";
import {
  type MapFormFieldToZodFn,
  mapFieldToZod,
  umbracoFormToZodSchema,
} from "../umbraco-form-to-zod";
import formDefinition from "./__fixtures__/UmbracoForm.fixture.json";

const defaultFieldTypes = Object.fromEntries(Object.entries(DefaultFieldType));
const defaultFieldKeys = Object.keys(
  defaultFieldTypes,
) as (keyof typeof defaultFieldTypes)[];

describe("mapFieldToZod", () => {
  describe("convert default fields to corresponding ZodType", () => {
    test.each(defaultFieldKeys.filter((key) => key !== "TitleAndDescription"))(
      "should convert %s to ZodType",
      (key) => {
        const zodType = mapFieldToZod({
          type: { id: defaultFieldTypes[key] },
        } as FormFieldDto);
        expect(zodType).toBeInstanceOf(z.ZodType);
      },
    );
  });

  describe("custom fields", () => {
    const customField = {
      type: {
        id: "CustomField",
        name: "CustomField",
      },
    } as FormFieldDto;

    test("should throw if custom mapping function is not provided and field is attempted to be converted", () => {
      expect(() => mapFieldToZod(customField)).toThrowError();
    });

    test("should throw if custom mapping function does not return a zodType", () => {
      const customMappingFunction = (() =>
        undefined) as unknown as MapFormFieldToZodFn;
      expect(() =>
        mapFieldToZod(customField, customMappingFunction),
      ).toThrowError();
    });

    test("should return zodType if custom mapping function is provided and correctly implemented", () => {
      const customMappingFunction: MapFormFieldToZodFn = () => z.string();
      const zodType = mapFieldToZod(customField, customMappingFunction);
      expect(zodType).toBeInstanceOf(z.ZodType);
    });

    test("should be optional if field defintion is marked as required: false", () => {
      const customMappingFunction: MapFormFieldToZodFn = () => z.string();
      const zodType = mapFieldToZod(
        { ...customField, required: false },
        customMappingFunction,
      );
      expect(zodType).toBeInstanceOf(z.ZodOptional);
    });
  });
});

describe("umbracoFormToZod", () => {
  test.skip("should convert form definition to zod schema", () => {
    const schema = umbracoFormToZodSchema(formDefinition as FormDto);
    expect(schema).toBeInstanceOf(z.ZodType);
    expect(schema).toMatchSnapshot();
  });
  test("should make required fields from schema optional if dependant", () => {
    const formData = {
      name: "test",
      email: "test@test.com",
      comment: "test",
      date: "02-02-2024",
      country: "gb",
      favouriteColour: ["red"],
      dataConsent: true,
      tickToAddMoreInfo: true,
      moreInfo: "",
    };
    const schema = umbracoFormToZodSchema(formDefinition as FormDto);
    const fields = getAllFields(formDefinition as FormDto);
    const moreInfoField = fields.find((field) => field?.alias === "moreInfo");
    expect(moreInfoField?.condition?.actionType).toBe("Show");
    expect(moreInfoField?.condition?.logicType).toBe("All");
    expect(moreInfoField?.condition?.rules).toMatchInlineSnapshot(`
      [
        {
          "field": "6ce0cf78-5102-47c1-85c6-9530d9e9c6a6",
          "operator": "Is",
          "value": "on",
        },
      ]
    `);

    const parsedSchema = schema.safeParse(formData);

    expect(parsedSchema.error?.issues).toMatchInlineSnapshot(`
      [
        {
          "code": "invalid_type",
          "fatal": true,
          "message": "Please provide a value for More info",
          "path": [
            "moreInfo",
          ],
        },
      ]
    `);

    expect(
      schema.safeParse({ ...formData, tickToAddMoreInfo: false }).error?.issues,
    ).toBe(undefined);
  });
});
