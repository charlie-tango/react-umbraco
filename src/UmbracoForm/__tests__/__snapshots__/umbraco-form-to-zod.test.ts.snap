// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`umbracoFormToZod > should convert form definition to zod schema 1`] = `
ZodObject {
  "_cached": null,
  "_def": {
    "catchall": ZodNever {
      "_def": {
        "typeName": "ZodNever",
      },
      "and": [Function],
      "array": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
    },
    "shape": [Function],
    "typeName": "ZodObject",
    "unknownKeys": "strip",
  },
  "and": [Function],
  "array": [Function],
  "augment": [Function],
  "brand": [Function],
  "catch": [Function],
  "default": [Function],
  "describe": [Function],
  "isNullable": [Function],
  "isOptional": [Function],
  "nonstrict": [Function],
  "nullable": [Function],
  "nullish": [Function],
  "optional": [Function],
  "or": [Function],
  "parse": [Function],
  "parseAsync": [Function],
  "pipe": [Function],
  "promise": [Function],
  "readonly": [Function],
  "refine": [Function],
  "refinement": [Function],
  "safeParse": [Function],
  "safeParseAsync": [Function],
  "spa": [Function],
  "superRefine": [Function],
  "transform": [Function],
}
`;
